qualtran/bloqs/block_encoding/linear_combination.py

```diff
@@ -0,0 +1,351 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import cast, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+from attrs import evolve, field, frozen, validators
+
+from qualtran import (
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    BoundedQUInt,
+    QAny,
+    Register,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran._infra.bloq import DecomposeTypeError
+from qualtran.bloqs.block_encoding import BlockEncoding
+from qualtran.bloqs.block_encoding.lcu_block_encoding import BlackBoxPrepare, BlackBoxSelect
+from qualtran.bloqs.block_encoding.lcu_select_and_prepare import PrepareOracle
+from qualtran.bloqs.block_encoding.phase import Phase
+from qualtran.bloqs.bookkeeping.auto_partition import AutoPartition, Unused
+from qualtran.bloqs.bookkeeping.partition import Partition
+from qualtran.linalg.lcu_util import preprocess_probabilities_for_reversible_sampling
+from qualtran.symbolics import smax, ssum, SymbolicFloat, SymbolicInt
+from qualtran.symbolics.types import is_symbolic
+
+
+@frozen
+class LinearCombination(BlockEncoding):
+    r"""Linear combination of a sequence of block encodings.
+
+    Builds the block encoding $B[\lambda_1 U_1 + \lambda_2 U_2 + \cdots + \lambda_n U_n]$ given
+    block encodings $B[U_1], \ldots, B[U_n]$ and coefficients $\lambda_i \in \mathbb{R}$.
+
+    When each $B[U_i]$ is a $(\alpha_i, a_i, \epsilon_i)$-block encoding of $U_i$, we have that
+    $B[\lambda_1 U_1 + \cdots + \lambda_n U_n]$ is a $(\sum_i \lvert\lambda_i\rvert\alpha_i, \lceil
+    \log_2 n \rceil + \max_i a_i, (\sum_i \lvert\lambda_i\rvert)\max_i \epsilon_i)$-block encoding
+    of $\lambda_1 U_1 + \cdots + \lambda_n U_n$.
+
+    Under the hood, this bloq uses LCU Prepare and Select oracles to build the block encoding.
+    These oracles will be automatically instantiated if not specified by the user.
+
+    Args:
```

Comment:
the way `attrs` data classes work is that if you define the attribute with a preceeding underscore, it will be "private" and users shouldn't be accessing that attribute on the class. But in the `__init__` constructor, the leading underscore is stripped from the argument name. When we document these constructor arguments (which is what we're doing here), lets remove the preceding underscore